js/io.ts

```diff
@@ -159,3 +161,31 @@ export function toAsyncIterator(r: Reader): AsyncIterableIterator<Uint8Array> {
     }
   };
 }
+
+/** Read lines asynchronously from `r`.
+ *
+ *      for await (const line of Deno.lines(reader)) {
+ *          console.log(line);
+ *      }
+ */
+export async function* lines(r: Reader): AsyncIterableIterator<string> {
+  const decoder = new TextDecoder();
+  let scratch = "";
+
+  for await (const chunk of toAsyncIterator(r)) {
```

Comment:
Ideally we could import code from deno_std here. The correct way to solve this is to use BufReader in std/io/bufio.ts
https://github.com/denoland/deno_std/blob/master/io/bufio.ts#L220

In order to use that code, we need to change //js/ to use extensions in the imports... (cc @kitsonk )

Reply:
Yeah, that is complicated (using extensions). I was planning to tackle it in the medium term. Essentially we have to replace rollup.

Reply:
Maybe we should move bufio into //js then (problem is that it will likely bring all of std/io with it as dependencies)

Reply:
Or maybe "lines" should just be in deno_std instead of core..

Reply:
It looks like `BufReader` provides a nice way to solve `lines` :+1:

Reading line by line is such a common operation that IMHO it should be `Deno.lines` instead of living in deno_std. 

It's greatly illustrated by @ry's example. This is super simple:
```bashsession
cat /etc/passwd |  deno eval " \
  for await (let line of Deno.lines(Deno.stdin) { \
    console.log(line.split(':')[0]) }"
```

This is not:
```bashsession
cat /etc/passwd |  deno eval " \
  import { lines } from "https://deno.land/x/io/mod.ts; \
  for await (let line of Deno.lines(Deno.stdin) { \
    console.log(line.split(':')[0]) }"
```

Reply:
@bartlomieju I agree it would be great to have built in to the CLI... but I really want to ensure that we use the proper bufio infrastructure to solve these problems rather than having a bunch of one-off algorithms for buffering. So either we need to first be able to import deno_std code into the CLI (ideal) or move bufio from deno_std into the main repo (less ideal)

Closing this one for now. Thanks @bartlomieju !